### ISSUE TITLE
Accepted old state should not be chanced during transient

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#10563]
## Rationale
I hope everybody agrees on the title. Currently this function https://github.com/idaholab/moose/blob/devel/framework/src/materials/MaterialPropertyStorage.C#L300
opens a hole, which developers could use to change the old state maybe unintentionally. I know nothing prevent developers from like changing old solution vectors, putting extra state (e.g. counter of linear residual evaluations) in kernel, material, etc. evaluations, but this one is pretty subtle.

## Description
I noticed this during debugging while I added another residual L2 norm evaluation at the end of `Transient::solveStep()`, which you would think changes nothing but breaks `materials/stateful_prop.stateful_copy`, which is using this. Just to open for a discussion.

## Impact
Could impact modules or applications using this capability.

**Comment by permcody** [comment]
Yeah, I can see your argument. At the very least copying the old and older states should be controlled by the object. I'm going to take an educated guess in Ben's case. With XFEM, this was indeed intentional since he is copying state from a parent element into children split elements so he's want the history there too. For other cases though, it might not make sense. I'd be fine if you added a parameter to this API to make the old/older state copying part optional unless you have something else in mind.

**Comment by YaqiWang** [comment]
This sounds related with mesh adaptation. I am not using this feature so here just to warn you what could happen.

**Comment by YaqiWang** [comment]
Some of the tests are good examples but some are not. The developers could follow a bad example. In some of the test code we should comment 'never follow this, it is used for testing only'.

**Comment by bwspenc** [comment]
@permcody is correct that this was intentional for XFEM.

That test broke because the copy is happening on TIMESTEP_END, and your change must be making the TIMESTEP_END stuff get called twice, which I imagine could break a lot of other things as well. I don't care too much about that test, other than that it is a way to test the material copy stuff outside of XFEM. It would be easy enough to hack MaterialCopyUserObject to only do that copy once per timestep, but I don't know if that's really the right thing to do.

**Comment by YaqiWang** [comment]
We must be doing similar things with mesh adaptation during transient. Feel free if you think this can be closed. I just think changing accepted old solution must be handled carefully, otherwise we could shoot our own foot easily.
